#  ////////*                                              
#         *///////////                                    
# //////////*   ////////////                              
#           ,/////////////////                            
# /////////////////////////////    ./////                 
#            //////////////////   ////// .///             
#     *///////////////////////// .//////.     ///         
#                 //////////////.  */////,*////  .//      
#                   /////////////.     ,/////  ,///* //   
#                     /////////////       /////      ,/*, 
#                      .//////////////.  ///////          
#                         //////////////////////          
#                             ,/////,*/////////           
#                               //////////////            
#                            ///////////////              
#                           /////////////                 
#                         ,/////////*                     
#                    .////////.                           

# This SnakeMake file contains the rule to run hiflye (flye with hifi parameters)
# Project: longread simulation study
# Description: perform de novo assembly on the simulated reads generated by PBSIM3 followed by ccs to generate HiFi consensus reads. 
#           The datasets have varying the readlength, sequencing depth and error rate
# Author: Paula Uittenbogaard
# Date: 13-05-2025

rule flye:
    input: 
        f"{ccs_dir}/{prefix}.fastq.gz"
    output:
        f"{hiflye_dir}/assembly.fasta",
        f"{hiflye_dir}/assembly_graph.gfa",
        f"{hiflye_dir}/assembly_graph.gv",
        f"{hiflye_dir}/flye.log"
    params:
        outdir = f"{hiflye_dir}"
    shell:
        """
        flye --pacbio-hifi {input} \
             --out-dir {params.outdir}
        """

rule rename_output_flye:
    input:
        fasta = f"{hiflye_dir}/assembly.fasta",
        gfa = f"{hiflye_dir}/assembly_graph.gfa",
        gv = f"{hiflye_dir}/assembly_graph.gv",
        log = f"{hiflye_dir}/flye.log"
    output:
        fasta = f"{hiflye_dir}/{prefix}.contigs.fasta",
        gfa = f"{hiflye_dir}/{prefix}.graph.gfa",
        gv = f"{hiflye_dir}/{prefix}.graph.gv",
        log = f"{hiflye_dir}/{prefix}.log"
    shell:
        """
        mv {input.fasta} {output.fasta}
        mv {input.gfa} {output.gfa}
        mv {input.gv} {output.gv}
        mv {input.log} {output.log}
        """