#  ////////*                                              
#         *///////////                                    
# //////////*   ////////////                              
#           ,/////////////////                            
# /////////////////////////////    ./////                 
#            //////////////////   ////// .///             
#     *///////////////////////// .//////.     ///         
#                 //////////////.  */////,*////  .//      
#                   /////////////.     ,/////  ,///* //   
#                     /////////////       /////      ,/*, 
#                      .//////////////.  ///////          
#                         //////////////////////          
#                             ,/////,*/////////           
#                               //////////////            
#                            ///////////////              
#                           /////////////                 
#                         ,/////////*                     
#                    .////////.                           

# This SnakeMake file contains the rule to run hicanu
# Project: longread simulation study
# Description: perform de novo assembly on the simulated reads generated by PBSIM3 followed by ccs to generate HiFi consensus reads. 
#           The datasets have varying the readlength, sequencing depth and error rate
# Author: Paula Uittenbogaard
# Date: 06-05-2025

rule hicanu:
    input: 
        f"{ccs_dir}/{prefix}.fastq.gz"
    resources:
        mem_mb = 5000
    threads: 10
    output:
        f"{hicanu_dir}/{prefix}.contigs.fasta"
    params:
        prefix = f"{prefix}",
        outdir = f"{hicanu_dir}"
    shell:
        """
        canu -p {params.prefix} \
             -d {params.outdir} \
             useGrid=false \
             genomeSize=1m \
             minInputCoverage=8 \
             stopOnLowCoverage=8 \
             -pacbio-hifi {input}
        """