{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is used to plot the metrics of the reads that are produced in the simulation study. The goal is to simulate reads with different characteristics (depth, readlength, accuracy) and produce assemblies. The read datasets will be characterized by:\n",
    "- Number of reads \n",
    "- Readlength distribution\n",
    "- Accuracy distribution\n",
    "-"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import pyplot as plt \n",
    "import numpy as np \n",
    "import bionumpy as bnp\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import os\n",
    "from os import environ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We want to plot:\n",
    "# - per-read accuracy\n",
    "# - readlength distribution\n",
    "# PBSIM3 does not output the accuracy in the fastq file so we will have to use the alignment of reads to the reference\n",
    "# In the snakemake pipeline minimap2 was used to align the reads back to their reference, resulting in a bam file.\n",
    "# Here htsbox samview is used to translate that back to a paf file \n",
    "# For paf file format see: https://cran.r-project.org/web/packages/pafr/vignettes/Introduction_to_pafr.html\n",
    "# This approach of using the alignment to calculate accuracy is also used by Ono-2022 (PBSIM3 paper): \"Accuracy [...] was computed from alignments between the reads and their reference genome\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract the fields we need from the .paf file\n",
    "# See also https://cran.r-project.org/web/packages/pafr/vignettes/Introduction_to_pafr.html\n",
    "# This is working for the paf files created by htsbox samview! (For paf files created by minimap2 directly, length is len(fields[9]) and nm is int(fields[11].split(':')[2])\n",
    "def parse_paf(file_path):\n",
    "    alignments = []\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            if line[0] == 'S':    \n",
    "                fields = line.strip().split('\\t')\n",
    "                #Check if this is not an entry that gives \"*\" as sequence\n",
    "                if len(fields[9])>1:\n",
    "                    alignment = {\n",
    "                        'query_name': fields[0],\n",
    "                        'query_length': int(fields[1]),\n",
    "                        'num_matches': int(fields[9]),\n",
    "                        'alen': int(fields[10])\n",
    "                        #'num_mismatches': int(fields[12].split(':')[2])\n",
    "                    }\n",
    "                alignments.append(alignment)\n",
    "    #for some reason there are duplicates (due to secondary alignments?) so we need to remove these\n",
    "    unique_alignments = [dict(t) for t in {tuple(d.items()) for d in alignments}]\n",
    "    return unique_alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "### ONT make paf file ###\n",
    "!find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/ONT\" -name *.fq.gz > ont_reads_filepaths.txt\n",
    "with open(\"ont_reads_filepaths.txt\") as filepaths:\n",
    "    for line in filepaths:\n",
    "        line = line.rstrip()\n",
    "        bam_line = line.replace(\".fq.gz\",\".bam\")\n",
    "        paf_line = line.replace(\".fq.gz\",\".paf\")\n",
    "        environ['bam_line'] = bam_line\n",
    "        environ['paf_line'] = paf_line\n",
    "        !htsbox samview -p \"$bam_line\" > \"$paf_line\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "!find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/ONT\" -name *.paf > ont_paf_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "### ONT plots ###\n",
    "# Read in all the paf files and plot their readlength and accuracy distributions\n",
    "# For the definition of BLAST identity (PAF file column 10 divided by 11), see https://epi2me.nanoporetech.com/quality-scores/ and https://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity\n",
    "!find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/ONT\" -name *.paf > ont_paf_files.txt\n",
    "with open(\"ont_paf_files.txt\") as pafs:\n",
    "    for paf in pafs:\n",
    "        paf = paf.rstrip()\n",
    "        alignments = parse_paf(paf)\n",
    "        all_accuracies=[]\n",
    "        all_lengths = []\n",
    "        ac_fig_filepath = paf.replace(\"_0001.paf\",\".blast_identity.png\")\n",
    "        rl_fig_filepath = paf.replace(\"_0001.paf\",\".readlengths.png\")\n",
    "        qc_line = paf.replace(\"_0001.paf\",\".accuracy_summary.txt\")\n",
    "        for alignment in alignments:\n",
    "            accuracy = alignment['num_matches']/alignment['alen']\n",
    "            all_accuracies.append(accuracy)\n",
    "            all_lengths.append(alignment['query_length'])\n",
    "        if \"rl10000\" in paf:\n",
    "            colours = 'green'\n",
    "        elif \"rl25000\" in paf:\n",
    "            colours = 'red'\n",
    "        elif \"rl50000\" in paf:\n",
    "            colours = 'blue'\n",
    "        ##Plot accuracies\n",
    "        plt.figure()\n",
    "        sns.histplot(all_accuracies,color=colours,edgecolor='white');\n",
    "        plt.title(\"Per-read accuracy distribution\")\n",
    "        plt.xlabel(\"Accuracy\")\n",
    "        plt.ylabel(\"Count\")\n",
    "        plt.savefig(ac_fig_filepath)\n",
    "        plt.close()\n",
    "        ##Plot readlengths\n",
    "        plt.figure()\n",
    "        sns.histplot(all_lengths,bins=50,color=colours,edgecolor='white');\n",
    "        plt.title(\"Readlength distribution\")\n",
    "        plt.xlabel(\"Readlength\")\n",
    "        plt.ylabel(\"Count\")\n",
    "        plt.xlim(0,100000)\n",
    "        plt.savefig(rl_fig_filepath)\n",
    "        plt.close()\n",
    "        median_accuracy = np.median(all_accuracies)\n",
    "        mean_accuracy = np.mean(all_accuracies)\n",
    "        sd_accuracy = np.std(all_accuracies)\n",
    "        with open(qc_line,'w') as qc_file:\n",
    "            print('Median accuracy',median_accuracy,'\\n','Mean accuracy',mean_accuracy,'\\n','sd',sd_accuracy,file=qc_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next up are the HiFi simulations (RSII produced by PBSIM3 then followed by CCS), plot readlength and accuracy distribution for each dataset in their own path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### HIFI make paf file ###\n",
    "!find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/HiFi\" -name *.fastq.gz > hifi_reads_filepaths.txt\n",
    "with open(\"hifi_reads_filepaths.txt\") as filepaths:\n",
    "    for line in filepaths:\n",
    "        line = line.rstrip()\n",
    "        bam_line = line.replace(\".fastq.gz\",\".bam\")\n",
    "        paf_line = line.replace(\".fastq.gz\",\".paf\")\n",
    "        environ['bam_line'] = bam_line\n",
    "        environ['paf_line'] = paf_line\n",
    "        !htsbox samview -p \"$bam_line\" > \"$paf_line\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### HIFI plots ###\n",
    "# Read in all the paf files and plot their readlength and accuracy distributions\n",
    "# For the definition of BLAST identity (PAF file column 10 divided by 11), see https://epi2me.nanoporetech.com/quality-scores/ and https://lh3.github.io/2018/11/25/on-the-definition-of-sequence-identity\n",
    "!find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/HiFi/ccs\" -name *.paf > hifi_paf_files.txt\n",
    "with open(\"hifi_paf_files.txt\") as pafs:\n",
    "    for paf in pafs:\n",
    "        paf = paf.rstrip()\n",
    "        alignments = parse_paf(paf)\n",
    "        all_accuracies=[]\n",
    "        all_lengths = []\n",
    "        ac_fig_filepath = paf.replace(\".paf\",\".blast_identity.png\")\n",
    "        rl_fig_filepath = paf.replace(\".paf\",\".readlenghts.png\")\n",
    "        qc_line = paf.replace(\".paf\",\".accuracy_summary.txt\")\n",
    "        for alignment in alignments:\n",
    "            accuracy = alignment['num_matches']/alignment['alen']\n",
    "            all_accuracies.append(accuracy)\n",
    "            all_lengths.append(alignment['query_length'])\n",
    "        if \"rl10000\" in paf:\n",
    "            colours = 'green'\n",
    "        elif \"rl17500\" in paf:\n",
    "            colours = 'red'\n",
    "        elif \"rl25000\" in paf:\n",
    "            colours = 'blue'\n",
    "        ## Plot accuracy\n",
    "        plt.figure()\n",
    "        sns.histplot(all_accuracies,color=colours,edgecolor='white');\n",
    "        plt.title(\"Per-read accuracy distribution\")\n",
    "        plt.xlabel(\"Accuracy\")\n",
    "        plt.ylabel(\"Count\")\n",
    "        plt.savefig(ac_fig_filepath)\n",
    "        plt.close()\n",
    "        ## Plot readlengths\n",
    "        plt.figure()\n",
    "        sns.histplot(all_lengths,bins=50,color=colours,edgecolor='white');\n",
    "        plt.title(\"Readlength distribution\")\n",
    "        plt.xlabel(\"Readlength\")\n",
    "        plt.ylabel(\"Count\")\n",
    "        plt.xlim(0,100000)\n",
    "        plt.savefig(rl_fig_filepath)\n",
    "        plt.close()\n",
    "        ## Write summary file of accuracies\n",
    "        median_accuracy = np.median(all_accuracies)\n",
    "        mean_accuracy = np.mean(all_accuracies)\n",
    "        sd_accuracy = np.std(all_accuracies)\n",
    "        with open(qc_line,'w') as qc_file:\n",
    "            print('Median accuracy',median_accuracy,'\\n','Mean accuracy',mean_accuracy,'\\n','sd',sd_accuracy,file=qc_file)\n",
    "!rm hifi_paf_files.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/HiFi\" -name *.fastq > hifi_unzipped_reads_filepaths.txt\n",
    "# with open(\"hifi_unzipped_reads_filepaths.txt\") as filepaths:\n",
    "#     for unzipped_line in filepaths:\n",
    "#         unzipped_line = unzipped_line.rstrip()\n",
    "#         zipped_line = unzipped_line.replace(\".fastq\",\".fastq.gz\")\n",
    "#         environ['unzipped'] = unzipped_line\n",
    "#         environ['zipped'] = zipped_line\n",
    "#         !gzip -c \"$unzipped\" > \"$zipped\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### ONT READLENGTHS FROM FASTA (not used anymore because this can be done directly from the paf file, which we need for accuracy distribution anyways)\n",
    "# Find all the paths to the simulated reads by PBSIM3 and generate a .readlengths.txt file that contains the counts of all readlengths\n",
    "# !find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/ONT\" -name *.fq.gz > ont_reads_filepaths.txt\n",
    "# with open(\"ont_reads_filepaths.txt\") as filepaths:\n",
    "#     for line in filepaths:\n",
    "#         line = line.rstrip()\n",
    "#         unzipped_line = line.replace(\".gz\",\"\")\n",
    "#         readlengths_filename = line.replace(\"_0001.fq.gz\",\".readlengths.txt\")\n",
    "#         environ['line'] = line\n",
    "#         environ['readlengths_filename'] = readlengths_filename\n",
    "#         environ['unzipped_line'] = unzipped_line\n",
    "#         !gunzip -dkf \"$line\" \n",
    "# #See https://www.biostars.org/p/72433/\n",
    "#         !awk 'NR % 4 ==2 {lengths[length($0)]++} END {for (len in lengths) print len, lengths[len]}' \"$unzipped_line\" > \"$readlengths_filename\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### ONT READLENGTHS\n",
    "# For all the readlengths.txt files generated in the cell above, plot the readlength distribution and save in their own path.=\n",
    "# See also https://bioinformatics.stackexchange.com/questions/45/read-length-distribution-from-fasta-file\n",
    "# !find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/ONT\" -name *.readlengths.txt > readlength_distributions.txt\n",
    "# with open(\"readlength_distributions.txt\") as distributions:\n",
    "#     for line in distributions:\n",
    "#         line = line.rstrip()\n",
    "#         rl_fig_filepath = line.replace(\".txt\",\".png\")\n",
    "#         length_distribution = pd.read_csv(line, delimiter=' ', header=None)\n",
    "#         if \"rl10000\" in line:\n",
    "#             colours = 'green'\n",
    "#         elif \"rl25000\" in line:\n",
    "#             colours = 'red'\n",
    "#         elif \"rl50000\" in line:\n",
    "#             colours = 'blue'\n",
    "#         plt.figure()\n",
    "#         sns.histplot(length_distribution[0],bins=50,color=colours,edgecolor='white');\n",
    "#         plt.title(\"Readlength distribution\")\n",
    "#         plt.xlabel(\"Readlength\")\n",
    "#         plt.ylabel(\"Count\")\n",
    "#         plt.xlim(0,100000)\n",
    "#         plt.savefig(rl_fig_filepath)\n",
    "#         plt.close()\n",
    "# !rm readlength_distributions.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### HIFI READLENGTHS\n",
    "# # Find all the paths to the simulated reads by PBSIM3 and generate a .readlengths.txt file that contains the counts of all readlengths\n",
    "# !find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/HiFi\" -name *.fastq.gz > hifi_reads_filepaths.txt\n",
    "# with open(\"hifi_reads_filepaths.txt\") as filepaths:\n",
    "#     for line in filepaths:\n",
    "#         line = line.rstrip()\n",
    "#         unzipped_line = line.replace(\".gz\",\"\")\n",
    "#         readlengths_filename = line.replace(\".fastq.gz\",\".readlengths.txt\")\n",
    "#         environ['line'] = line\n",
    "#         environ['readlengths_filename'] = readlengths_filename\n",
    "#         environ['unzipped_line'] = unzipped_line\n",
    "#         !gunzip -dkf \"$line\" \n",
    "# #See https://www.biostars.org/p/72433/\n",
    "#         !awk 'NR % 4 ==2 {lengths[length($0)]++} END {for (len in lengths) print len, lengths[len]}' \"$unzipped_line\" > \"$readlengths_filename\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ### HIFI READLENGTHS\n",
    "# # For all the readlengths.txt files generated in the cell above, plot the readlength distribution and save in their own path.=\n",
    "# # See also https://bioinformatics.stackexchange.com/questions/45/read-length-distribution-from-fasta-file\n",
    "# !find \"/home/uitte01p/experimental/test_for_sim_study/simulation_output/HiFi\" -name *.readlengths.txt > hifi_readlength_distributions.txt\n",
    "# with open(\"hifi_readlength_distributions.txt\") as distributions:\n",
    "#     for line in distributions:\n",
    "#         line = line.rstrip()\n",
    "#         rl_fig_filepath = line.replace(\".txt\",\".png\")\n",
    "#         length_distribution = pd.read_csv(line, delimiter=' ', header=None)\n",
    "#         if \"rl10000\" in line:\n",
    "#             colours = 'green'\n",
    "#         elif \"rl17500\" in line:\n",
    "#             colours = 'red'\n",
    "#         elif \"rl25000\" in line:\n",
    "#             colours = 'blue'\n",
    "#         plt.figure()\n",
    "#         sns.histplot(length_distribution[0],bins=50,color=colours,edgecolor='white');\n",
    "#         plt.title(\"Readlength distribution\")\n",
    "#         plt.xlabel(\"Readlength\")\n",
    "#         plt.ylabel(\"Count\")\n",
    "#         plt.xlim(0,100000)\n",
    "#         plt.savefig(rl_fig_filepath)\n",
    "#         plt.close()\n",
    "# !rm hifi_readlength_distributions.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "simulation_study",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
